### ISSUE TITLE
Log flooding with new proc plugin

### ISSUE THREAD

**Issue by TuurDutoit** [netdata/netdata#4945]
##### Bug report summary
My Netdata logs are flooded with the following message:

    netdata ERROR : PLUGIN[proc] : Core 1 found but no more than 1 cores were expected.

This is running in a Proxmox container which is allocated 1 core. The Proxmox environment itself runs on an Intel Xeon E5-2683 v3 (14 cores, 28 threads - sadly not my own hardware...).  
My `/proc/cpuinfo` correctly reports I have 1 core. However, `/proc/schedstat` contains info about `cpu0` through `cpu55` (56 total, or 4x14 for some weird reason. Maybe there's 2 CPUs, I'm not sure), as does my `/sys/devices/system/cpu` directory.
The proc plugin detects this discrepancy and consequently throws an error at:
https://github.com/netdata/netdata/blob/25dc3d83e9ccfc2dd000e2a08a84be94d684a498/collectors/proc.plugin/proc_stat.c#L317-L320

This plugin was moved from Python to C a few days ago in #4635
##### OS / Environment
netdata 1.12.0-rc0-11-g4e879a2c
Debian 9.6 (stretch)
Linux 4.13.13-3-pve
Running as container in Proxmox 5.2-10
##### Component Name
`proc` plugin
##### Steps To Reproduce
Run Netdata in a container that isn't allocated all cores; I assume that's what leads to the discrepancy between `/proc/cpuinfo` (reporting the number of allocated CPU cores) and `/proc/schedstat` and `/sys/devices/system/cpu` (reporting the total amount of hardware cores) which ultimately leads to the logging of an error message every second.
##### Expected behavior
No error is logged as this discrepancy is to be expected in a virtualized environment.

**Comment by cakrit** [comment]
@vlvkobal and @ktsaou I commented out the error line to at least prevent the log flood until a proper fix is delivered.

**Comment by cakrit** [comment]
@TuurDutoit please do an update to at least get the workaround that will prevent the log flood. 

So that we can do the fix properly, could you please send us your /proc/cpuinfo and the line from /proc/schedstat that matches the processor shown in /proc/cpuinfo? From what I see on my pc, I expect we'll see something like e.g. processor : 25 in /proc/cpuinfo and cpu25 in /proc/schedstat.

**Comment by cakrit** [comment]
Well, reading up [here](https://github.com/moby/moby/issues/20770), it sounds like `/proc/cpuinfo` won't give us what we need and that we'll have to parse `/sys/fs/cgroup/cpuset/cpuset.cpus`. Can you `cat` that one too please?

**Comment by TuurDutoit** [comment]
`cat /proc/cpuinfo`:
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 63
model name	: Intel(R) Xeon(R) CPU E5-2683 v3 @ 2.00GHz
stepping	: 2
microcode	: 0x2d
cpu MHz		: 1995.487
cache size	: 35840 KB
physical id	: 0
siblings	: 28
core id		: 0
cpu cores	: 14
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 15
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts
bugs		: cpu_insecure
bogomips	: 3990.97
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```

`cat /proc/schedstat`:
```
version 15
timestamp 10410754698
cpu0 0 0 0 0 0 0 466736388923771 554040485424644 10755922460
domain0 000000,10000001 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
domain1 555555,55555555 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
domain2 ffffff,ffffffff 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
cpu1 0 0 0 0 0 0 879929491698795 108656000876520 3902799006
domain0 000000,20000002 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
domain1 aaaaaa,aaaaaaaa 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
domain2 ffffff,ffffffff 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
cpu2 0 0 0 0 0 0 546246338009031 764422908310402 4970711020
domain0 000000,40000004 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
domain1 555555,55555555 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
domain2 ffffff,ffffffff 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
cpu3 0 0 0 0 0 0 721588286401835 19337006925229 3656020567
domain0 000000,80000008 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
domain1 aaaaaa,aaaaaaaa 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
domain2 ffffff,ffffffff 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
(goes on like this up to cpu55)
```

`ls -1 /sys/devices/system/cpu`:
```
cpu0
cpu1
cpu10
cpu11
cpu12
cpu13
cpu14
cpu15
cpu16
cpu17
cpu18
cpu19
cpu2
cpu20
cpu21
cpu22
cpu23
cpu24
cpu25
cpu26
cpu27
cpu28
cpu29
cpu3
cpu30
cpu31
cpu32
cpu33
cpu34
cpu35
cpu36
cpu37
cpu38
cpu39
cpu4
cpu40
cpu41
cpu42
cpu43
cpu44
cpu45
cpu46
cpu47
cpu48
cpu49
cpu5
cpu50
cpu51
cpu52
cpu53
cpu54
cpu55
cpu6
cpu7
cpu8
cpu9
cpufreq
cpuidle
hotplug
intel_pstate
isolated
kernel_max
microcode
modalias
offline
online
possible
power
present
uevent
```

`cat /sys/fs/cgroup/cpuset/cpuset.cpus`:
```
0
```

**Comment by TuurDutoit** [comment]
I don't currently have another container, though...
If need be, I could ask for another one; the guy who manages the servers actually recommended Netdata to me!

**Comment by cakrit** [comment]
Ok, this is interesting. I started a docker container on my manjaro with `--cpuset-cpus="3-4"`. In my case, both `/proc/schedstat` and `/proc/cpuinfo` showed all 12 cores. It was only `/sys/fs/cgroup/cpuset/cpuset.cpus` that showed `3-4`. So I wasn't able to replicate your situation and it seems to be OS-specific. It changes nothing regarding the next step of waiting for some input from our experts, but at least it means that the bug doesn't affect everyone.

**Comment by ktsaou** [comment]
If I understand this correctly, the problem is that only one cpu is available, but instead of it being named `cpu0`, it is called `cpu1`.

So, the problem is that the plugin expects a sequentially numbered set of cpus, but they could be random.

Right?

**Comment by cakrit** [comment]
> If I understand this correctly, the problem is that only one cpu is available, but instead of it being named cpu0, it is called cpu1.

Not really, it's that sometimes `/prov/cpuinfo` will show us fewer cpus that what `/proc/schedstat` shows. We haven't been able to replicate it yet with a cpu in `/prov/cpuinfo` that is not the equivalent of `cpu0`, so we're not quite sure what it would show if it was e.g. `cpu11`. If in this case `/prov/cpuinfo` shows `processor: 11`, then we can definitely match that to to `/proc/schedstat`. If not, we'll need to look at `/sys/fs/cgroup/cpuset/cpuset.cpus`. The key question is whether we actually want containers handled by proc, or if we want to leave cpuidle for these cases to become part of the cgroups collector. I'm leaning towards the second option.

**Comment by ktsaou** [comment]
> The key question is whether we actually want containers handled by proc, or if we want to leave cpuidle for these cases to become part of the cgroups collector. I'm leaning towards the second option.

cpu states are not exposed per cgroup. So making `cpuidle` become part of the cgroups collector is meaningless. Am I missing something?

To my understanding the `cpuidle` module needs a small refactor. Instead of an array of `struct per_core_cpuidle_chart`, it should have a linked list of them and lookup the right structure for each cpu found.

This way, we won't care if the kernel files match or not. We will always present the information that is available.

**Comment by socomsystems** [comment]
@Ferroin @vlvkobal this wouldn't be related to the DOS prevention triggers I've been experiencing when proc has been flooding my error logs due to /proc/schedstat not existing.  It crippled my slave nodes > master collector setup today after executing fresh installs on all my Debian 8 Jessie slave nodes.  Had to proc = no [plugins] and HUP netdata to stop backlogs, and net alarm inundations.

```
2018-12-20 12:55:38: netdata ERROR : PLUGIN[proc] : PROCFILE: Cannot open file '/proc/schedstat' (errno 2, No such file or directory)
2018-12-20 12:55:39: netdata LOG FLOOD PROTECTION too many logs (201 logs in 125 seconds, threshold is set to 200 logs in 1200 seconds). Preventing more logs from process 'netdata' for 1075 seconds.
```
Oddly enough its only impacted vps on Vultr as far can tell thus far.  However, I did just check one of my Debian 8 Jessie instances on DO and schedstat seems to be ghosted as well.

 apologies for the cross-posts: [https://github.com/netdata/netdata/issues/4920](https://github.com/netdata/netdata/issues/4920)

**Comment by cakrit** [comment]
@socomsystems : Missing `proc/schedstat` and the required fix have been discussed in #4994. @vlvkobal will work on it soon.

**Comment by vlvkobal** [comment]
@TuurDutoit The issue should be fixed in #5065. Please test it.
